packages/core/admin/admin/src/core/apis/rbac.ts

```diff
@@ -0,0 +1,58 @@
+import { Location } from 'react-router-dom';
+
+import type { Permission, User } from '../../features/Auth';
+
+interface RBACContext extends Pick<Location, 'pathname' | 'search'> {
+  /**
+   * The current user.
+   */
+  user?: User;
+  /**
+   * The permissions of the current user.
+   */
+  permissions: Permission[];
+}
+
+interface RBACMiddleware {
+  (
+    ctx: RBACContext
```

Comment:
the main idea was to make the context immutable, similar to how redux middlewares work

Reply:
Ok, just looked up how redux middleware work, since they do it that way it makes sense to do it differently for front-end. I'll switch to thinking about it like that.